File: orion-platform-service/src/services/session/SessionRepository.ts
import { DatabasePool } from '../database';
/**
 * SessionRepository - Database layer for Session operations
 */

export interface Session {
  id: string;
  user_id: string;
  tenant_id: string;
  token: string;
  expires_at: Date;
  created_at: Date;
}

export class SessionRepository {
  constructor(private pool: DatabasePool) {}

  async create(userId: string, tenantId: string, token: string, expiresAt: Date): Promise<Session> {
    const result = await this.pool.query(
      'INSERT INTO sessions (user_id, tenant_id, token, expires_at) VALUES ($1, $2, $3, $4) RETURNING *',
      [userId, tenantId, token, expiresAt]
    );
    return result.rows[0];
  }

  async findByToken(token: string): Promise<Session | null> {
    const result = await this.pool.query(
      'SELECT * FROM sessions WHERE token = $1 AND expires_at > NOW()',
      [token]
    );
    return result.rows[0] || null;
  }

  async revoke(token: string): Promise<boolean> {
    const result = await this.pool.query('DELETE FROM sessions WHERE token = $1', [token]);
    return result.rowCount > 0;
  }

  async cleanup(): Promise<number> {
    const result = await this.pool.query('DELETE FROM sessions WHERE expires_at < NOW()');
    return result.rowCount;
  }

  async findByUser(userId: string, tenantId?: string): Promise<Session[]> {
    if (tenantId) {
      const result = await this.pool.query(
        'SELECT * FROM sessions WHERE user_id = $1 AND tenant_id = $2 AND expires_at > NOW() ORDER BY created_at DESC',
        [userId, tenantId]
      );
      return result.rows;
    }
    const result = await this.pool.query(
      'SELECT * FROM sessions WHERE user_id = $1 AND expires_at > NOW() ORDER BY created_at DESC',
      [userId]
    );
    return result.rows;
  }

  async refresh(token: string, extendHours: number = 24): Promise<Session | null> {
    const expiresAt = new Date(Date.now() + extendHours * 60 * 60 * 1000);
    const result = await this.pool.query(
      'UPDATE sessions SET expires_at = $2 WHERE token = $1 AND expires_at > NOW() RETURNING *',
      [token, expiresAt]
    );
    return result.rows[0] || null;
  }
}

File: orion-auth-svc/internal/repository/session_repository.go
package repository

import (
	"context"
	"fmt"

	"orion/auth-svc/internal/models"
	"orion/go-common/pkg/database"

	"github.com/jmoiron/sqlx"
)

// SessionRepository provides data access for session entities.
type SessionRepository struct {
	database.BaseRepository
}

func NewSessionRepository(db *sqlx.DB) *SessionRepository {
	return &SessionRepository{
		BaseRepository: database.NewBaseRepository(db),
	}
}

func (r *SessionRepository) Create(ctx context.Context, session *models.Session) error {
	query := `
		INSERT INTO sessions (user_id, tenant_id, token, ip_address, user_agent, expires_at)
		VALUES ($1, $2, $3, $4, $5, $6) RETURNING id, created_at
	`
	err := r.DB().QueryRowContext(ctx, query,
		session.UserID, session.TenantID, session.Token,
		session.IP, session.UserAgent, session.ExpiresAt,
	).Scan(&session.ID, &session.CreatedAt)
	return err
}

func (r *SessionRepository) GetByToken(ctx context.Context, token string) (*models.Session, error) {
	var session models.Session
	query := `SELECT id, user_id, tenant_id, token, ip_address, user_agent, expires_at, created_at FROM sessions WHERE token = $1`
	err := r.DB().GetContext(ctx, &session, query, token)
	if err != nil {
		return nil, fmt.Errorf("session not found: %w", err)
	}
	return &session, nil
}

func (r *SessionRepository) GetByUserID(ctx context.Context, userID string) ([]models.Session, error) {
	var sessions []models.Session
	query := `SELECT id, user_id, tenant_id, token, ip_address, user_agent, expires_at, created_at FROM sessions WHERE user_id = $1 AND expires_at > now() ORDER BY created_at DESC`
	err := r.DB().SelectContext(ctx, &sessions, query, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get sessions: %w", err)
	}
	return sessions, nil
}

func (r *SessionRepository) Delete(ctx context.Context, id string) error {
	_, err := r.DB().ExecContext(ctx, "DELETE FROM sessions WHERE id = $1", id)
	return err
}

func (r *SessionRepository) DeleteByUserID(ctx context.Context, userID string) error {
	_, err := r.DB().ExecContext(ctx, "DELETE FROM sessions WHERE user_id = $1", userID)
	return err
}

func (r *SessionRepository) CleanupExpired(ctx context.Context) (int64, error) {
	result, err := r.DB().ExecContext(ctx, "DELETE FROM sessions WHERE expires_at < now()")
	if err != nil {
		return 0, err
	}
	return result.RowsAffected()
}
